### ISSUE TITLE
clarify dev priorities

### ISSUE THREAD

**Issue by maxheld83** [maxheld83/ghactions#117]
### TLDR

I'd like to first work on getting basic CI/CD functionality done, tested and documented.
First todos are:

- figure out the caching / docker image situation #110 #106 
- figure out a good unit/integration testing regime #114 #116 #22 

Then move on  to the more fancy/interesting stuff.

----

I'm still trying to avoid too much planning-overhead, but in revisiting the old issues, some clarification of dev priorities might be helpful to ensure that:

- this aligns with what the (tidyverse) team needs
- this makes (some) technical sense.

The following are tradeoffs, where I'd always suggest to prioritize the **former over the latter**.

### pkg dev vs. other R projects

As @jimhester and I discussed earlier, the priority for now is **pkg dev**, especially the needs of the tidyverse and r-lib teams.

There's still a bunch of wild non-pkg-dev related issues floating around from earlier this year, but those are all backlogged for now.

Possible/minor exceptions might be maintenance on the RMarkdown action and/or using it as a test case (because it's pretty easy).

### Simple CI vs. fancy stuff

Simple CI would encompass reproducing the base Travis CI features (build, install, check).
Fancier stuff includes programmatic commits (`devtools::document()`) and friends as well as programmatic code review (`goodpractice::goodpractice()` and many more).
For a more detailed breakdown, see #2.

The fancier stuff might have interesting potential to improve the teams productivity, and relies on some capabilities that *only* (?) github actions (= full/easy github api access) has.
So this is definitely worth trying.

But I'm **leaning towards doing simple CI first**, because:

- it's easier
- there are some foundational issues to be resolved first (see below).

AFAIK there's currently only *one* R package that relies on GitHub actions for CI, and that's *this* one. The setup isn't very good (bad dockerfile #106, #69, etc.), and it's not documented anywhere. So I'm leaning to working on that first.

### Singleplatform vs Matrix-Testing/Multiplatform

AFAIK, github actions right now isn't very well suited to testing multiple OSes or R-versions (there's no support for matrix testing or any of that).
I'd suggest to accept this as a limitation, and **to *only* do single platform**, say debian-release, for now (as per #106, will ask around which r-hub image is best).

Later on, we can add multiplatform testing by adding actions to also test on r-hub #71 or even ms azure pipelines #114, but these are relatively slow/expensive and might not even be necessary for every commit and at every project stage / branch, etc.

### Actions vs. Templating

My medium-term (end of summer) goal would be to have a `usethis::use_ghactions()` function, which sets up a complete, thoroughly tested, `.github/main.workflow` which should work for most r packages #41.
Perhaps, there might be additional tidyverse/r-lib defaults à la `usethis::use_tidy_ci()` #115 and maybe #87.

This would save everyone a lot of time, because you don't want to set up a dozend `.github/main.workflow`s manually, and it's easy to make mistakes.

This requires some templating functions, which already exist in this package for RMarkdown projects. They write out action blocks and combine them into workflow blocks.
This wouldn't be an interface *on top of* the (visually editable) `main.workflow` HCL (as `tic.r` would be in #56), but a fairly dumb templater with reasonable defaults.. 
For some special/edge cases, there'll always be the visual editor (which I think is pretty good/better than `tic.r` to reason about).

Still, this requires solid actions to begin with, so I'd **prioritize building actions** first (that is, a `Dockerfile`, some `entrypoint.sh` bash with maybe some github api calls and solid testing and documentation (see below).

In the extreme, this could put actions in their own repos (that's what github currently requires on their marketplace), but I guess the overhead would be too high. ~~or, submodules 😨~~ 

There *is* a slight complication here when actions are broken down to essentially *one* R command each. Arguably, they should be broken down in such small pieces, because there are no travis-like stages in github actions to distinguish between stages.

It's possible that some of these actions are *so* similar (they basically just differ in some `Rscript -e foo::bar()` call) that it might be better *not* to build separate actions, but just separate templating calls for the *same* action. But I'm not sure.
This also depends on the caching situation #110 #48 and is currently a bit unclear to me.

### Testing vs. ...

I'd like to start out by building very thorough integration #22 and unit tests #116 #14  tests for the simple CI first.
I was really slowed down by testing the actions manually, and it'd also just give me piece of mind / make it easier to debug things.

I think medium-term, a solid testing suite that tests ghactions actions/workflows against, say, some known boilerplate pkg #22 might also be an asset in helping users (pkg devs) debug their builds etc, if there is a baseline to compare it to.

I'm not 100% sure on how to best/elegantly do this, but I think it would be worth the effort to get this right.

### image-off-container-registry vs. custom dockerfile

This one is documented in #106 and also weirdly depends on the caching situation #110 #48.
More details in #106.
I'd like to figure this out first, and am heavily leaning **towards using images-off-container-registries** (from r-hub) for pkg dev.

**Comment by maxheld83** [comment]
will do as suggested, but caching will take a backseat for now.
Just focus on simple steps and testing.

**Comment by maxheld83** [comment]
also we decided to spin out actions into separate repos for now.
